src/librustc_mir/transform/const_prop.rs

```diff
@@ -560,6 +545,32 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
                     return None;
                 }
             }
+
+            Rvalue::Aggregate(_, operands) if operands.len() == 0 => {
+                // FIXME(wesleywiser): const eval will turn this into a `const Scalar(<ZST>)` that
+                // `SimplifyLocals` doesn't know it can remove.
+                return None;
+            }
+
+            Rvalue::Aggregate(box AggregateKind::Adt(_, variant_idx, _, _, _), _) => {
+                match &place_layout.variants {
+                    Variants::Single { index } => {
+                        if index != variant_idx {
+                            trace!("skipping uninhabited variant {:?}", rvalue);
+                            return None;
```

Comment:
I think what happens here (and I think this needs a comment explaining that and the `trace!` should be clarified) is that we are in unreachable code. Like, this code is UB if it ever gets executed. const-prop does not generally test for UB, but for some reason it has to detect this UB here... or so?

If you made the Miri engine `write_discriminant` cause a UB error instead of ICE'ing for bad discriminants, could you then remove this code here again? I think that would be much cleaner.